compio-driver/src/asyncify.rs

```diff
@@ -0,0 +1,86 @@
+use std::{
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc,
+    },
+    time::Duration,
+};
+
+use crossbeam_channel::{bounded, Receiver, Sender, TrySendError};
+
+type BoxClosure = Box<dyn FnOnce() + Send>;
+
+struct CounterGuard(Arc<AtomicUsize>);
+
+impl Drop for CounterGuard {
+    fn drop(&mut self) {
+        self.0.fetch_sub(1, Ordering::AcqRel);
+    }
+}
+
+fn worker(
+    receiver: Receiver<BoxClosure>,
+    counter: Arc<AtomicUsize>,
+    timeout: Duration,
+) -> impl FnOnce() {
+    move || {
+        counter.fetch_add(1, Ordering::AcqRel);
+        let _guard = CounterGuard(counter);
+        while let Ok(f) = receiver.recv_timeout(timeout) {
+            f();
+        }
+    }
+}
+
+/// A thread pool to perform blocking operations in other threads.
+///
+/// ## Platform specific
+/// * io-uring: the driver doesn't user this thread pool.
+#[derive(Debug)]
+pub struct AsyncifyPool {
+    sender: Sender<BoxClosure>,
+    receiver: Receiver<BoxClosure>,
+    counter: Arc<AtomicUsize>,
+    thread_limit: usize,
+    recv_timeout: Duration,
+}
+
+impl AsyncifyPool {
+    /// Create [`AsyncifyPool`] with thread number limit and channel receive
+    /// timeout.
+    pub fn new(thread_limit: usize, recv_timeout: Duration) -> Self {
+        let (sender, receiver) = bounded(0);
+        Self {
+            sender,
+            receiver,
+            counter: Arc::new(AtomicUsize::new(0)),
+            thread_limit,
+            recv_timeout,
+        }
+    }
+
+    /// Send a closure to another thread. Usually the user should not use it.
+    pub fn dispatch(&self, f: impl FnOnce() + Send + 'static) -> bool {
```

Comment:
Return the closure if not spawned. This may require manual unsafe downcast (`*const dyn FnOnce as *const F`)
```rust
pub fn dispatch<F: FnOnce() + Send + 'static>(&self, f: F) -> Result<(), F> { .. }
```